lms/djangoapps/grades/tests/test_models.py

```diff
@@ -0,0 +1,216 @@
+            vblocks.blocks = blocks
+
+
+@ddt.ddt
+class PersistentSubsectionGradeTest(GradesModelTestCase):
+    """
+    Test the PersistentSubsectionGrade model.
+    """
+    def setUp(self):
+        super(PersistentSubsectionGradeTest, self).setUp()
+        self.usage_key = BlockUsageLocator(
+            course_key=self.course_key,
+            block_type='subsection',
+            block_id='subsection_12345',
+        )
+        self.params = {
+            "user_id": "student12345",
+            "usage_key": self.usage_key,
+            "course_version": "deadbeef",
+            "subtree_edited_date": "2016-08-01 18:53:24.354741",
+            "earned_all": 6,
+            "possible_all": 12,
+            "earned_graded": 6,
+            "possible_graded": 8,
+            "visible_blocks": [self.record_a, self.record_b],
+        }
+
+    def test_create(self):
+        """
+        Tests model creation, and confirms error when trying to recreate model.
+        """
+        created_grade = PersistentSubsectionGrade.create(**self.params)
+        read_grade = PersistentSubsectionGrade.read(user_id=self.params["user_id"], usage_key=self.params["usage_key"])
+        self.assertEqual(created_grade, read_grade)
+        with self.assertRaises(IntegrityError):
+            created_grade = PersistentSubsectionGrade.create(**self.params)
+
+    @ddt.data(
+        ("missing", KeyError),
+        ("invalid", AttributeError),
+    )
+    @ddt.unpack
+    def test_create_bad_params(self, ddt_name, error):
+        """
+        Confirms create will fail if params are missing or invalid.
+        """
+        if ddt_name == "missing":
+            del self.params["course_version"]
+        elif ddt_name == "invalid":
+            self.params["usage_key"] = "usage_keys_are_strings_now"
+        with self.assertRaises(error):
+            _ = PersistentSubsectionGrade.create(**self.params)
+
+    def test_update(self):
+        """
+        Tests model update, and confirms error when updating a nonexistent model.
+        """
+        with self.assertRaises(PersistentSubsectionGrade.DoesNotExist):
+            PersistentSubsectionGrade.update(**self.params)
+        _ = PersistentSubsectionGrade.create(**self.params)
+        self.params['earned_all'] = 12
+        self.params['earned_graded'] = 8
+        PersistentSubsectionGrade.update(**self.params)
+        read_grade = PersistentSubsectionGrade.read(user_id=self.params["user_id"], usage_key=self.params["usage_key"])
+        self.assertEqual(read_grade.earned_all, 12)
+        self.assertEqual(read_grade.earned_graded, 8)
+
+    @ddt.data(True, False)
+    def test_save(self, already_created):
+        if already_created:
+            _ = PersistentSubsectionGrade.create(**self.params)
+            mock_fn_called = "update"
+            mock_fn_uncalled = "create"
+        else:
+            mock_fn_called = "update"
+            mock_fn_uncalled = "create"
```

Comment:
mock_fn_called and mock_fn_uncalled are set to the same values in both branches.  Either this is a typo, or it should be outside the if/else block.